pkg/addonfactory/chartaddonfactory/chartaddonagent.go

```diff
@@ -0,0 +1,132 @@
+package chartaddonfactory
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/fatih/structs"
+	"helm.sh/helm/v3/pkg/chart"
+	"helm.sh/helm/v3/pkg/chartutil"
+	"helm.sh/helm/v3/pkg/engine"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/klog/v2"
+	"open-cluster-management.io/addon-framework/pkg/agent"
+	addonapiv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+)
+
+// the annotation name of customized chart values
+const annotationValuesName string = "addon.open-cluster-management.io/values"
+
+// the reserved fields for template config
+const (
+	clusterName           string = "clusterName"
+	addonInstallNamespace string = "addonInstallNamespace"
+	hubKubeConfigSecret   string = "hubKubeConfigSecret"
+)
+
+type ChartAgentAddon struct {
+	decoder           runtime.Decoder
+	chart             *chart.Chart
+	values            interface{}
+	agentAddonOptions agent.AgentAddonOptions
+}
+
+func newChartAgentAddon(
+	scheme *runtime.Scheme,
+	chart *chart.Chart,
+	values interface{},
+	agentAddonOptions agent.AgentAddonOptions) *ChartAgentAddon {
+	return &ChartAgentAddon{
+		decoder:           serializer.NewCodecFactory(scheme).UniversalDeserializer(),
+		chart:             chart,
+		values:            values,
+		agentAddonOptions: agentAddonOptions}
+}
+
+func (a *ChartAgentAddon) Manifests(
+	cluster *clusterv1.ManagedCluster,
+	addon *addonapiv1alpha1.ManagedClusterAddOn) ([]runtime.Object, error) {
+	var objects []runtime.Object
+
+	values, err := a.getValues(cluster, addon)
+	if err != nil {
+		return objects, err
+	}
+
+	helmEngine := engine.Engine{
+		Strict:   true,
+		LintMode: false,
+	}
+
+	templates, err := helmEngine.Render(a.chart, values)
+	if err != nil {
+		return objects, err
+	}
+	for _, data := range templates {
+		if len(data) == 0 {
+			continue
+		}
+
+		object, _, err := a.decoder.Decode([]byte(data), nil, nil)
+		if err != nil {
+			return nil, err
+		}
+		objects = append(objects, object)
+	}
+
+	return objects, nil
+}
+
+func (a *ChartAgentAddon) GetAgentAddonOptions() agent.AgentAddonOptions {
+	return a.agentAddonOptions
+}
+
+func (a *ChartAgentAddon) getValues(
+	cluster *clusterv1.ManagedCluster,
+	addon *addonapiv1alpha1.ManagedClusterAddOn) (map[string]interface{}, error) {
+	overrideValues := structs.Map(a.values)
+	overrideValues[clusterName] = cluster.GetName()
+
+	installNamespace := addon.Spec.InstallNamespace
+	if len(installNamespace) == 0 {
+		installNamespace = AddonDefaultInstallNamespace
+	}
+	overrideValues[addonInstallNamespace] = installNamespace
+	overrideValues[hubKubeConfigSecret] = fmt.Sprintf("%s-hub-kubeconfig", a.agentAddonOptions.AddonName)
```

Comment:
I would suggest having a separate func for builtin values. And hubKubeConfigSecret needs to be defined based on addon status

Reply:
use `getBuildInValues` to get all of build-in values. 
define `hubKubeConfigSecret ` when `a.agentAddonOptions.Registration != nil`